File: swift/Sources/Rezel/Grammars/SassGrammar.swift
import Foundation

// MARK: - Character constants

private let sassSpace: [Int] = [9, 10, 11, 12, 13, 32, 133, 160, 5760, 8192, 8193, 8194, 8195, 8196, 8197,
                                8198, 8199, 8200, 8201, 8202, 8232, 8233, 8239, 8287, 12288]

private let colon = 58
private let parenL = 40
private let underscore = 95
private let bracketL = 91
private let dash = 45
private let period = 46
private let hash = 35
private let percent = 37
private let braceL = 123
private let braceR = 125
private let slash = 47
private let asterisk = 42
private let newlineChar = 10
private let equals = 61
private let plus = 43
private let ampersand = 38

// MARK: - Helpers

private func isAlpha(_ ch: Int) -> Bool {
	(ch >= 65 && ch <= 90) || (ch >= 97 && ch <= 122) || ch >= 161
}

private func isDigit(_ ch: Int) -> Bool {
	ch >= 48 && ch <= 57
}

private func sassStartOfComment(_ input: InputStream) -> Bool {
	if input.next != slash { return false }
	let next = input.peek(1)
	return next == slash || next == asterisk
}

private func sassIndented(_ stack: Stack) -> Bool {
	let keys = Array(stack.parser.dialects.keys)
	if let idx = keys.firstIndex(of: "indented") {
		return stack.dialectEnabled(idx)
	}
	return false
}

// MARK: - Context

private class SassIndentLevel {
	let parent: SassIndentLevel?
	let depth: Int
	let hashValue: Int

	init(parent: SassIndentLevel?, depth: Int) {
		self.parent = parent
		self.depth = depth
		let parentHash = parent?.hashValue ?? 0
		hashValue = parentHash &+ (parentHash << 8) &+ depth &+ (depth << 4)
	}
}

private nonisolated(unsafe) let sassTopIndent = SassIndentLevel(parent: nil, depth: 0)

// MARK: - Context tracker

private nonisolated(unsafe) let sassTrackIndent = ContextTracker(
	start: sassTopIndent as Any,
	shift: { context, term, stack, input in
		guard let cx = context as? SassIndentLevel else { return context }
		let name = stack.parser.termNames?[term] ?? ""
		if name == "indent" {
			return SassIndentLevel(parent: cx, depth: stack.pos - input.pos) as Any
		}
		if name == "dedent" {
			return (cx.parent ?? sassTopIndent) as Any
		}
		return context
	},
	hash: { context in (context as? SassIndentLevel)?.hashValue ?? 0 }
)

// MARK: - External tokenizers

private func makeSassExternalTokenizer(name: String, terms: [String: Int]) -> TokenizerProtocol {
	if name == "spaces" {
		let eofT = terms["eof"]!
		let blankLineStartT = terms["blankLineStart"]!
		let newlineT = terms["newline"]!
		let whitespaceT = terms["whitespace"]!
		return ExternalTokenizer({ input, stack in
			if sassIndented(stack) {
				if input.next < 0, stack.canShift(eofT) {
					input.acceptToken(eofT)
				} else {
					let prev = input.peek(-1)
					if prev == newlineChar || prev < 0, stack.canShift(blankLineStartT) {
						var spaces = 0
						while input.next != newlineChar && sassSpace.contains(input.next) {
							input.advance()
							spaces += 1
						}
						if input.next == newlineChar || sassStartOfComment(input) {
							input.acceptToken(blankLineStartT, endOffset: -spaces)
						} else if spaces > 0 {
							input.acceptToken(whitespaceT)
						}
					} else if input.next == newlineChar {
						input.acceptToken(newlineT, endOffset: 1)
					} else if sassSpace.contains(input.next) {
						input.advance()
						while input.next != newlineChar, sassSpace.contains(input.next) {
							input.advance()
						}
						input.acceptToken(whitespaceT)
					}
				}
			} else {
				var length = 0
				while sassSpace.contains(input.next) {
					input.advance()
					length += 1
				}
				if length > 0 {
					input.acceptToken(whitespaceT)
				}
			}
		}, contextual: true)
	}

	if name == "comments" {
		let lineCommentT = terms["LineComment"]!
		let commentT = terms["Comment"]!
		return ExternalTokenizer { input, stack in
			guard sassStartOfComment(input) else { return }
			input.advance()
			if sassIndented(stack) {
				var indentedComment = -1
				var off = 1
				while true {
					let prev = input.peek(-off - 1)
					if prev == newlineChar || prev < 0 {
						indentedComment = off + 1
						break
					} else if !sassSpace.contains(prev) {
						break
					}
					off += 1
				}
				if indentedComment > -1 {
					let block = input.next == asterisk
					var end = 0
					input.advance()
					while input.next >= 0 {
						if input.next == newlineChar {
							input.advance()
							var indented = 0
							while input.next != newlineChar, sassSpace.contains(input.next) {
								indented += 1
								input.advance()
							}
							if indented < indentedComment {
								end = -indented - 1
								break
							}
						} else if block, input.next == asterisk, input.peek(1) == slash {
							end = 2
							break
						} else {
							input.advance()
						}
					}
					input.acceptToken(block ? commentT : lineCommentT, endOffset: end)
					return
				}
			}
			if input.next == slash {
				while input.next != newlineChar, input.next >= 0 {
					input.advance()
				}
				input.acceptToken(lineCommentT)
			} else {
				input.advance()
				while input.next >= 0 {
					let next = input.next
					input.advance()
					if next == asterisk, input.next == slash {
						input.advance()
						break
					}
				}
				input.acceptToken(commentT)
			}
		}
	}

	if name == "indentedMixins" {
		let indentedMixinT = terms["IndentedMixin"]!
		let indentedIncludeT = terms["IndentedInclude"]!
		return ExternalTokenizer { input, stack in
			if input.next == plus || input.next == equals, sassIndented(stack) {
				input.acceptToken(input.next == equals ? indentedMixinT : indentedIncludeT, endOffset: 1)
			}
		}
	}

	if name == "indentation" {
		let dedentT = terms["dedent"]!
		let indentT = terms["indent"]!
		return ExternalTokenizer { input, stack in
			guard sassIndented(stack) else { return }
			let cDepth = (stack.context as? SassIndentLevel)?.depth ?? 0
			if input.next < 0, cDepth > 0 {
				input.acceptToken(dedentT)
				return
			}
			let prev = input.peek(-1)
			if prev == newlineChar {
				var depth = 0
				while input.next != newlineChar, sassSpace.contains(input.next) {
					input.advance()
					depth += 1
				}
				if depth != cDepth,
				   input.next != newlineChar, !sassStartOfComment(input)
				{
					if depth < cDepth {
						input.acceptToken(dedentT, endOffset: -depth)
					} else {
						input.acceptToken(indentT)
					}
				}
			}
		}
	}

	if name == "identifiers" {
		let interpolationStartT = terms["InterpolationStart"]!
		let variableNameT = terms["VariableName"]!
		let queryIdentifierT = terms["queryIdentifier"]!
		let calleeT = terms["callee"]!
		let identifierT = terms["identifier"]!
		return ExternalTokenizer { input, stack in
			var inside = false
			var dashes = 0
			var i = 0
			while true {
				let next = input.next
				if isAlpha(next) || next == dash || next == underscore || (inside && isDigit(next)) {
					if !inside, next != dash || i > 0 { inside = true }
					if dashes == i, next == dash { dashes += 1 }
					input.advance()
				} else if next == hash, input.peek(1) == braceL {
					input.acceptToken(interpolationStartT, endOffset: 2)
					break
				} else {
					if inside {
						let token: Int
						if dashes == 2, stack.canShift(variableNameT) {
							token = variableNameT
						} else if stack.canShift(queryIdentifierT) {
							token = queryIdentifierT
						} else if next == parenL {
							token = calleeT
						} else {
							token = identifierT
						}
						input.acceptToken(token)
					}
					break
				}
				i += 1
			}
		}
	}

	if name == "interpolationEnd" {
		let interpolationContinueT = terms["InterpolationContinue"]!
		let interpolationEndT = terms["InterpolationEnd"]!
		return ExternalTokenizer { input, _ in
			if input.next == braceR {
				input.advance()
				while isAlpha(input.next) || input.next == dash || input.next == underscore || isDigit(input.next) {
					input.advance()
				}
				if input.next == hash, input.peek(1) == braceL {
					input.acceptToken(interpolationContinueT, endOffset: 2)
				} else {
					input.acceptToken(interpolationEndT)
				}
			}
		}
	}

	if name == "descendant" {
		let descendantOpT = terms["descendantOp"]!
		return ExternalTokenizer { input, _ in
			if sassSpace.contains(input.peek(-1)) {
				let next = input.next
				if isAlpha(next) || next == underscore || next == hash || next == period ||
					next == bracketL || (next == colon && isAlpha(input.peek(1))) ||
					next == dash || next == ampersand || next == asterisk
				{
					input.acceptToken(descendantOpT)
				}
			}
		}
	}

	if name == "unitToken" {
		let unitT = terms["Unit"]!
		return ExternalTokenizer { input, _ in
			if !sassSpace.contains(input.peek(-1)) {
				let next = input.next
				if next == percent {
					input.advance()
					input.acceptToken(unitT)
				}
				if isAlpha(next) {
					repeat {
						input.advance()
					} while isAlpha(input.next) || isDigit(input.next)
					input.acceptToken(unitT)
				}
			}
		}
	}

	fatalError("Unknown Sass external tokenizer: \(name)")
}

// MARK: - Highlighting

private nonisolated(unsafe) let constantMod = hlTags["constant"] as! (Tag) -> Tag
private nonisolated(unsafe) let specialMod = hlTags["special"] as! (Tag) -> Tag

nonisolated(unsafe) let sassHighlighting = styleTags([
	"AtKeyword import charset namespace keyframes media supports include mixin use forward extend at-root": hlTags["definitionKeyword"] as Any,
	"Keyword selector": hlKeyword,
	"ControlKeyword": hlTags["controlKeyword"] as Any,
	"NamespaceName": hlTags["namespace"] as Any,
	"KeyframeName": hlTags["labelName"] as Any,
	"KeyframeRangeName": hlTags["operatorKeyword"] as Any,
	"TagName": hlTags["tagName"] as Any,
	"ClassName Suffix": hlTags["className"] as Any,
	"PseudoClassName": constantMod(hlTags["className"] as! Tag),
	"IdName": hlTags["labelName"] as Any,
	"FeatureName PropertyName": hlPropertyName,
	"AttributeName": hlTags["attributeName"] as Any,
	"NumberLiteral": hlNumber,
	"KeywordQuery": hlKeyword,
	"UnaryQueryOp": hlTags["operatorKeyword"] as Any,
	"CallTag ValueName": hlTags["atom"] as Any,
	"VariableName": hlTags["variableName"] as Any,
	"SassVariableName": specialMod(hlTags["variableName"] as! Tag),
	"Callee": hlTags["operatorKeyword"] as Any,
	"Unit": hlTags["unit"] as Any,
	"UniversalSelector NestingSelector IndentedMixin IndentedInclude": hlTags["definitionOperator"] as Any,
	"MatchOp": hlTags["compareOperator"] as Any,
	"ChildOp SiblingOp, LogicOp": hlTags["logicOperator"] as Any,
	"BinOp": hlTags["arithmeticOperator"] as Any,
	"Important Global Default": hlTags["modifier"] as Any,
	"Comment": hlTags["blockComment"] as Any,
	"LineComment": hlTags["lineComment"] as Any,
	"ColorLiteral": hlTags["color"] as Any,
	"ParenthesizedContent StringLiteral": hlString,
	"InterpolationStart InterpolationContinue InterpolationEnd": hlMeta,
	": \"...\"": hlPunctuation,
	"PseudoOp #": hlTags["derefOperator"] as Any,
	"; ,": hlTags["separator"] as Any,
	"( )": hlTags["paren"] as Any,
	"[ ]": hlTags["squareBracket"] as Any,
	"{ }": hlTags["brace"] as Any,
])

// MARK: - Parser

public nonisolated(unsafe) let sassParser: LRParser = {
	let externals: [String: TokenizerProtocol] = [
		"spaces": makeSassExternalTokenizer(name: "spaces", terms: SassParserData.termTable),
		"comments": makeSassExternalTokenizer(name: "comments", terms: SassParserData.termTable),
		"indentedMixins": makeSassExternalTokenizer(name: "indentedMixins", terms: SassParserData.termTable),
		"indentation": makeSassExternalTokenizer(name: "indentation", terms: SassParserData.termTable),
		"identifiers": makeSassExternalTokenizer(name: "identifiers", terms: SassParserData.termTable),
		"interpolationEnd": makeSassExternalTokenizer(name: "interpolationEnd", terms: SassParserData.termTable),
		"descendant": makeSassExternalTokenizer(name: "descendant", terms: SassParserData.termTable),
		"unitToken": makeSassExternalTokenizer(name: "unitToken", terms: SassParserData.termTable),
	]
	let spec = SassParserData.makeSpec(
		externals: externals,
		propSources: [sassHighlighting],
		context: sassTrackIndent
	)
	return LRParser(spec: spec)
}()

File: dotnet/Rezel/Grammars/SassGrammar.cs
using Rezel.Common;
using Rezel.Lr;
using Rezel.Highlight;

namespace Rezel.Grammars;

public static class SassGrammar
{
    private static readonly HashSet<int> Space =
    [
        9, 10, 11, 12, 13, 32, 133, 160, 5760, 8192, 8193, 8194, 8195, 8196, 8197,
        8198, 8199, 8200, 8201, 8202, 8232, 8233, 8239, 8287, 12288
    ];

    private const int Colon = 58;
    private const int ParenL = 40;
    private const int Underscore = 95;
    private const int BracketL = 91;
    private const int Dash = 45;
    private const int Period = 46;
    private const int Hash = 35;
    private const int Percent = 37;
    private const int BraceL = 123;
    private const int BraceR = 125;
    private const int Slash = 47;
    private const int Asterisk = 42;
    private const int NewlineChar = 10;
    private const int EqualsChar = 61;
    private const int Plus = 43;
    private const int And = 38;

    private static bool IsAlpha(int ch) => (ch >= 65 && ch <= 90) || (ch >= 97 && ch <= 122) || ch >= 161;
    private static bool IsDigit(int ch) => ch >= 48 && ch <= 57;

    private static bool StartOfComment(InputStream input)
    {
        return input.Next == Slash && ((input.Peek(1) == Slash) || (input.Peek(1) == Asterisk));
    }

    private class IndentLevel
    {
        public IndentLevel? Parent;
        public int Depth;
        public int Hash;

        public IndentLevel(IndentLevel? parent, int depth)
        {
            Parent = parent;
            Depth = depth;
            Hash = (parent != null ? parent.Hash + (parent.Hash << 8) : 0) + depth + (depth << 4);
        }
    }

    private static readonly IndentLevel TopIndent = new(null, 0);

    private static ContextTracker MakeSassTrackIndent()
    {
        return new ContextTracker(
            start: TopIndent,
            shift: (context, term, stack, input) =>
            {
                var ctx = (IndentLevel)context!;
                var name = stack.Parser.TermNames?.GetValueOrDefault(term) ?? "";
                if (name == "indent") return new IndentLevel(ctx, stack.Pos - input.Pos);
                if (name == "dedent") return ctx.Parent!;
                return ctx;
            },
            hash: (context) => ((IndentLevel)context!).Hash
        );
    }

    private static int? DialectIndex(LRParser parser, string name)
    {
        var keys = parser.Dialects.Keys.ToArray();
        var idx = Array.IndexOf(keys, name);
        return idx >= 0 ? idx : null;
    }

    private static ITokenizer MakeSassExternalTokenizer(string name, Dictionary<string, int> terms)
    {
        if (name == "spaces")
        {
            var eof = terms["eof"];
            var blankLineStart = terms["blankLineStart"];
            var newlineTerm = terms["newline"];
            var whitespace = terms["whitespace"];

            return new ExternalTokenizer((input, stack) =>
            {
                var indentedIdx = DialectIndex(stack.Parser, "indented");
                if (indentedIdx.HasValue && stack.DialectEnabled(indentedIdx.Value))
                {
                    if (input.Next < 0 && stack.CanShift(eof))
                    {
                        input.AcceptToken(eof);
                    }
                    else
                    {
                        var prev = input.Peek(-1);
                        if ((prev == NewlineChar || prev < 0) && stack.CanShift(blankLineStart))
                        {
                            var spaces = 0;
                            while (input.Next != NewlineChar && Space.Contains(input.Next)) { input.Advance(); spaces++; }
                            if (input.Next == NewlineChar || StartOfComment(input))
                                input.AcceptToken(blankLineStart, endOffset: -spaces);
                            else if (spaces > 0)
                                input.AcceptToken(whitespace);
                        }
                        else if (input.Next == NewlineChar)
                        {
                            input.AcceptToken(newlineTerm, endOffset: 1);
                        }
                        else if (Space.Contains(input.Next))
                        {
                            input.Advance();
                            while (input.Next != NewlineChar && Space.Contains(input.Next)) input.Advance();
                            input.AcceptToken(whitespace);
                        }
                    }
                }
                else
                {
                    var length = 0;
                    while (Space.Contains(input.Next))
                    {
                        input.Advance();
                        length++;
                    }
                    if (length > 0) input.AcceptToken(whitespace);
                }
            }, contextual: true);
        }

        if (name == "comments")
        {
            var LineComment = terms["LineComment"];
            var Comment = terms["Comment"];

            return new ExternalTokenizer((input, stack) =>
            {
                if (!StartOfComment(input)) return;
                input.Advance();
                var indentedIdx = DialectIndex(stack.Parser, "indented");
                if (indentedIdx.HasValue && stack.DialectEnabled(indentedIdx.Value))
                {
                    var indentedComment = -1;
                    for (var off = 1; ; off++)
                    {
                        var prev = input.Peek(-off - 1);
                        if (prev == NewlineChar || prev < 0)
                        {
                            indentedComment = off + 1;
                            break;
                        }
                        else if (!Space.Contains(prev))
                        {
                            break;
                        }
                    }
                    if (indentedComment > -1)
                    {
                        var block = input.Next == Asterisk;
                        var end = 0;
                        input.Advance();
                        while (input.Next >= 0)
                        {
                            if (input.Next == NewlineChar)
                            {
                                input.Advance();
                                var indented = 0;
                                while (input.Next != NewlineChar && Space.Contains(input.Next))
                                {
                                    indented++;
                                    input.Advance();
                                }
                                if (indented < indentedComment)
                                {
                                    end = -indented - 1;
                                    break;
                                }
                            }
                            else if (block && input.Next == Asterisk && input.Peek(1) == Slash)
                            {
                                end = 2;
                                break;
                            }
                            else
                            {
                                input.Advance();
                            }
                        }
                        input.AcceptToken(block ? Comment : LineComment, endOffset: end);
                        return;
                    }
                }
                if (input.Next == Slash)
                {
                    while (input.Next != NewlineChar && input.Next >= 0) input.Advance();
                    input.AcceptToken(LineComment);
                }
                else
                {
                    input.Advance();
                    while (input.Next >= 0)
                    {
                        var next = input.Next;
                        input.Advance();
                        if (next == Asterisk && input.Next == Slash)
                        {
                            input.Advance();
                            break;
                        }
                    }
                    input.AcceptToken(Comment);
                }
            });
        }

        if (name == "indentedMixins")
        {
            var IndentedMixin = terms["IndentedMixin"];
            var IndentedInclude = terms["IndentedInclude"];

            return new ExternalTokenizer((input, stack) =>
            {
                var indentedIdx = DialectIndex(stack.Parser, "indented");
                if (indentedIdx.HasValue && stack.DialectEnabled(indentedIdx.Value))
                {
                    if (input.Next == Plus || input.Next == EqualsChar)
                        input.AcceptToken(input.Next == EqualsChar ? IndentedMixin : IndentedInclude, endOffset: 1);
                }
            });
        }

        if (name == "indentation")
        {
            var dedent = terms["dedent"];
            var indent = terms["indent"];

            return new ExternalTokenizer((input, stack) =>
            {
                var indentedIdx = DialectIndex(stack.Parser, "indented");
                if (!indentedIdx.HasValue || !stack.DialectEnabled(indentedIdx.Value)) return;
                var ctx = (IndentLevel)stack.Context!;
                if (input.Next < 0 && ctx.Depth != 0)
                {
                    input.AcceptToken(dedent);
                    return;
                }
                var prev = input.Peek(-1);
                if (prev == NewlineChar)
                {
                    var depth = 0;
                    while (input.Next != NewlineChar && Space.Contains(input.Next))
                    {
                        input.Advance();
                        depth++;
                    }
                    if (depth != ctx.Depth &&
                        input.Next != NewlineChar && !StartOfComment(input))
                    {
                        if (depth < ctx.Depth) input.AcceptToken(dedent, endOffset: -depth);
                        else input.AcceptToken(indent);
                    }
                }
            });
        }

        if (name == "identifiers")
        {
            var identifier = terms["identifier"];
            var callee = terms["callee"];
            var varName = terms["VariableName"];
            var queryIdentifier = terms["queryIdentifier"];
            var InterpolationStart = terms["InterpolationStart"];

            return new ExternalTokenizer((input, stack) =>
            {
                var inside = false;
                var dashes = 0;
                for (var i = 0; ; i++)
                {
                    var next = input.Next;
                    if (IsAlpha(next) || next == Dash || next == Underscore || (inside && IsDigit(next)))
                    {
                        if (!inside && (next != Dash || i > 0)) inside = true;
                        if (dashes == i && next == Dash) dashes++;
                        input.Advance();
                    }
                    else if (next == Hash && input.Peek(1) == BraceL)
                    {
                        input.AcceptToken(InterpolationStart, endOffset: 2);
                        break;
                    }
                    else
                    {
                        if (inside)
                            input.AcceptToken(
                                dashes == 2 && stack.CanShift(varName) ? varName
                                : stack.CanShift(queryIdentifier) ? queryIdentifier
                                : next == ParenL ? callee
                                : identifier);
                        break;
                    }
                }
            });
        }

        if (name == "interpolationEnd")
        {
            var InterpolationEnd = terms["InterpolationEnd"];
            var InterpolationContinue = terms["InterpolationContinue"];

            return new ExternalTokenizer((input, _) =>
            {
                if (input.Next == BraceR)
                {
                    input.Advance();
                    while (IsAlpha(input.Next) || input.Next == Dash || input.Next == Underscore || IsDigit(input.Next))
                        input.Advance();
                    if (input.Next == Hash && input.Peek(1) == BraceL)
                        input.AcceptToken(InterpolationContinue, endOffset: 2);
                    else
                        input.AcceptToken(InterpolationEnd);
                }
            });
        }

        if (name == "descendant")
        {
            var descendantOp = terms["descendantOp"];
            return new ExternalTokenizer((input, _) =>
            {
                if (Space.Contains(input.Peek(-1)))
                {
                    var next = input.Next;
                    if (IsAlpha(next) || next == Underscore || next == Hash || next == Period ||
                        next == BracketL || (next == Colon && IsAlpha(input.Peek(1))) || next == Dash || next == And || next == Asterisk)
                        input.AcceptToken(descendantOp);
                }
            });
        }

        if (name == "unitToken")
        {
            var unit = terms["Unit"];
            return new ExternalTokenizer((input, _) =>
            {
                if (!Space.Contains(input.Peek(-1)))
                {
                    var next = input.Next;
                    if (next == Percent) { input.Advance(); input.AcceptToken(unit); }
                    if (IsAlpha(next))
                    {
                        do { input.Advance(); } while (IsAlpha(input.Next) || IsDigit(input.Next));
                        input.AcceptToken(unit);
                    }
                }
            });
        }

        throw new ArgumentException($"Unknown Sass external tokenizer: {name}");
    }

    private static readonly NodePropSource SassHighlighting = HighlightUtil.StyleTags(new Dictionary<string, object>
    {
        ["AtKeyword import charset namespace keyframes media supports include mixin use forward extend at-root"] = Tags.DefinitionKeyword,
        ["Keyword selector"] = Tags.Keyword,
        ["ControlKeyword"] = Tags.ControlKeyword,
        ["NamespaceName"] = Tags.Namespace,
        ["KeyframeName"] = Tags.LabelName,
        ["KeyframeRangeName"] = Tags.OperatorKeyword,
        ["TagName"] = Tags.TagName,
        ["ClassName Suffix"] = Tags.ClassName,
        ["PseudoClassName"] = Tags.Constant(Tags.ClassName),
        ["IdName"] = Tags.LabelName,
        ["FeatureName PropertyName"] = Tags.PropertyName,
        ["AttributeName"] = Tags.AttributeName,
        ["NumberLiteral"] = Tags.Number,
        ["KeywordQuery"] = Tags.Keyword,
        ["UnaryQueryOp"] = Tags.OperatorKeyword,
        ["CallTag ValueName"] = Tags.Atom,
        ["VariableName"] = Tags.VariableName,
        ["SassVariableName"] = Tags.Special(Tags.VariableName),
        ["Callee"] = Tags.OperatorKeyword,
        ["Unit"] = Tags.Unit,
        ["UniversalSelector NestingSelector IndentedMixin IndentedInclude"] = Tags.DefinitionOperator,
        ["MatchOp"] = Tags.CompareOperator,
        ["ChildOp SiblingOp, LogicOp"] = Tags.LogicOperator,
        ["BinOp"] = Tags.ArithmeticOperator,
        ["Important Global Default"] = Tags.ModifierTag,
        ["Comment"] = Tags.BlockComment,
        ["LineComment"] = Tags.LineComment,
        ["ColorLiteral"] = Tags.Color,
        ["ParenthesizedContent StringLiteral"] = Tags.String,
        ["InterpolationStart InterpolationContinue InterpolationEnd"] = Tags.Meta,
        [": \"...\""] = Tags.Punctuation,
        ["PseudoOp #"] = Tags.DerefOperator,
        ["; ,"] = Tags.Separator,
        ["( )"] = Tags.Paren,
        ["[ ]"] = Tags.SquareBracket,
        ["{ }"] = Tags.Brace,
    });

    private static LRParser? _parser;
    public static LRParser Parser => _parser ??= CreateParser();

    private static LRParser CreateParser()
    {
        var termTable = SassParserData.TermTable!;
        var externals = new Dictionary<string, ITokenizer>
        {
            ["indentation"] = MakeSassExternalTokenizer("indentation", termTable),
            ["descendant"] = MakeSassExternalTokenizer("descendant", termTable),
            ["interpolationEnd"] = MakeSassExternalTokenizer("interpolationEnd", termTable),
            ["unitToken"] = MakeSassExternalTokenizer("unitToken", termTable),
            ["identifiers"] = MakeSassExternalTokenizer("identifiers", termTable),
            ["spaces"] = MakeSassExternalTokenizer("spaces", termTable),
            ["comments"] = MakeSassExternalTokenizer("comments", termTable),
            ["indentedMixins"] = MakeSassExternalTokenizer("indentedMixins", termTable),
        };
        var spec = SassParserData.MakeSpec(
            externals: externals,
            propSources: [SassHighlighting],
            context: MakeSassTrackIndent()
        );
        return new LRParser(spec);
    }
}
